cpp/src/arrow/compute/light_array.h

```diff
@@ -0,0 +1,384 @@
+  Status ResizeVaryingLengthBuffer();
+  /// \brief The current length (in rows) of the array
+  int num_rows() const { return num_rows_; }
+  /// \brief A non-owning view into this array
+  KeyColumnArray column_array() const;
+  /// \brief A lightweight descriptor of the data held by this array
+  KeyColumnMetadata column_metadata() const {
+    return ColumnMetadataFromDataType(data_type_);
+  }
+  /// \brief Convert the data to an arrow::ArrayData
+  ///
+  /// This is a zero copy operation and the created ArrayData will reference the
+  /// buffers held by this instance.
+  std::shared_ptr<ArrayData> array_data() const;
+  /// \brief A raw pointer to the requested buffer
+  ///
+  /// If i is 0 then this returns the validity buffer
+  /// If i is 1 then this returns the buffer used for values (if this is a fixed
+  ///           length data type) or offsets (if this is a variable binary type)
+  /// If i is 2 then this returns the buffer used for variable length binary data
+  uint8_t* mutable_data(int i) {
+    return i == 0   ? non_null_buf_->mutable_data()
+           : i == 1 ? fixed_len_buf_->mutable_data()
+                    : var_len_buf_->mutable_data();
+  }
+
+ private:
+  static constexpr int64_t kNumPaddingBytes = 64;
+  int log_num_rows_min_;
+  std::shared_ptr<DataType> data_type_;
+  MemoryPool* pool_;
+  int num_rows_;
+  int num_rows_allocated_;
+  int var_len_buf_size_;
+  std::shared_ptr<ResizableBuffer> non_null_buf_;
+  std::shared_ptr<ResizableBuffer> fixed_len_buf_;
+  std::shared_ptr<ResizableBuffer> var_len_buf_;
+};
+
+/// \brief A builder to concatenate batches of data into a larger batch
+///
+/// Will only store num_rows_max() rows
+class ExecBatchBuilder {
+ public:
+  /// \brief Add rows from `source` into `target` column
+  ///
+  /// If `target` is uninitialized or cleared it will be initialized to use
+  /// the given pool.
+  static Status AppendSelected(const std::shared_ptr<ArrayData>& source,
+                               ResizableArrayData* target, int num_rows_to_append,
+                               const uint16_t* row_ids, MemoryPool* pool);
+
+  /// \brief Add nulls into `target` column
+  ///
+  /// If `target` is uninitialized or cleared it will be initialized to use
+  /// the given pool.
+  static Status AppendNulls(const std::shared_ptr<DataType>& type,
+                            ResizableArrayData& target, int num_rows_to_append,
+                            MemoryPool* pool);
+
+  /// \brief Add selected rows from `batch`
+  ///
+  /// If `col_ids` is null then `num_cols` should less than batch.num_values() and
+  /// the first `num_cols` columns of batch will be appended.
+  ///
+  /// All columns in `batch` must have array shape
+  Status AppendSelected(MemoryPool* pool, const ExecBatch& batch, int num_rows_to_append,
+                        const uint16_t* row_ids, int num_cols,
+                        const int* col_ids = NULLPTR);
```

Comment:
It's a little confusing that we have overloads that store `num_appended` and overloads that do not store `num_appended`.  The former are also able to exceed `num_rows_max()` I think.  What is the intended use case for the overloads that don't store `num_appended`?

Reply:
AppendSelected and AppendNulls variants that returned num_appended were wrappers on top of versions that did not. They can be removed as long as the caller is willing to make sure they don't exceed the batch size limit (there isn't anything incorrect about exceeding the limit at this point, but consumers of exec batches often assume that the rows can be indexed using 16-bit integers).

I removed variants that returned num_appended (they were not used in new hash join) and added returning error status in case of exceeding limit.